### PR TITLE
Subscribe to library changes

### PR DESCRIPTION
**Pull request by eliska-n**
I started with git provider. It appeared to be quite straightforward. But I would like to continue with at least the FileSystem provider before merging to prove it viable.

The user should be able to subscribe to the whole library (`library.subscribe()` without argument) or a specific path. There is one "subscribe" event for the whole library split among all registered providers. I realized I do not need to store and update changes in the library files - I just check whether the path of the file that has changed starts with the subscribed path. And I hope it is sufficient. It looks too easy to be true.

### CODE REVIEW

**Comment by ateska** [comment]
> `library.subscribe()` without argument

That is not needed.

**Comment by eliska-n** [comment]
`library.subscribe()` is equivalent to `library.subscribe("/")`. 

Yes, it could be automatic. Subscribe to the whole library without asking. It is simpler, but it would be overhead in the case when I don't need to track the changes, and specifically in the remote-control, I want to subscribe to the `/Site` path, I don't want to reload the microservice on every change in the library.

**Comment by ateska** [comment]
> `library.subscribe()` is equivalent to `library.subscribe("/")`.
> 
> Yes, it could be automatic. Subscribe to the whole library without asking. It is simpler, but it would be overhead in the case when I don't need to track the changes, and specifically in the remote-control, I want to subscribe to the `/Site` path, I don't want to reload the microservice on every change in the library.

You will hardly need subscription to the whole library, for this reason.

File: asab/library/providers/git.py [hunk 1]
```diff
@@ -72,6 +72,8 @@ def __init__(self, library, path):
 		self.ProactorService = self.App.get_service("asab.ProactorService")
 		self.PullLock = False
 
+		self.SubscribedPath = None
```
Comment (ateska) [inline asab/library/providers/git.py:75]: It should be a set/list ... and I suggest to consider having this in `abc`.

File: examples/library-subscribe.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+
+	async def on_library_ready(self, event_name, library=None):
+		items = await self.LibraryService.list("", recursive=True)
+		print("# Library\n")
+		for item in items:
+			print(" *", item)
+		print("\n===")
+
+		self.LibraryService.subscribe()
```
Comment (ateska) [inline examples/library-subscribe.py:35]: Use path here.
It is the example, people will copy this from here.

File: asab/library/providers/git.py [hunk 3]
```diff
@@ -72,6 +72,8 @@ def __init__(self, library, path):
 		self.ProactorService = self.App.get_service("asab.ProactorService")
 		self.PullLock = False
 
+		self.SubscribedPath = []
```
Comment (ateska) [inline asab/library/providers/git.py:75]: `self.SubscribedPath = set()`

File: asab/library/providers/git.py [hunk 4]
```diff
@@ -134,15 +136,33 @@ def fetch(self):
 
 
 	def pull(self):
-
 		new_commit_id = self.fetch()
 
+		# Before new head is set, check the diffs. If changes in subscribed directory occured, set `publish` flag.
+
+		publish = False
```
Comment (ateska) [inline asab/library/providers/git.py:143]: `pubchanges = []`

File: asab/library/providers/git.py [hunk 5]
```diff
@@ -134,15 +136,33 @@ def fetch(self):
...
+		for path in self.SubscribedPath:
+			for i in self.GitRepository.diff(self.GitRepository.head.target, new_commit_id).deltas:
+				if path == "/":
+					publish = True
```
Comment (ateska) [inline asab/library/providers/git.py:147]: `pubchanges.append(...)`

File: asab/library/providers/git.py [hunk 6]
```diff
@@ -134,15 +136,33 @@ def fetch(self):
...
 			return
 
+		# Reset HEAD
 		self.GitRepository.head.set_target(new_commit_id)
 		self.GitRepository.reset(new_commit_id, pygit2.GIT_RESET_HARD)
-		self.App.PubSub.publish("GitProviderUpdated!", self)
+
+		# Once reset of the head is finished, PubSub message about the change in the subsrcibed directory gets published.
+		if publish:
+			self.App.PubSub.publish("ASABLibrary.change!", self)
```
Comment (ateska) [inline asab/library/providers/git.py:162]: `self.App.PubSub.publish("ASABLibrary.change!", self, paths=pubchanges)`

**Review by ateska — CHANGES_REQUESTED** [review]

File: asab/library/providers/git.py (hunk 1 above)
Comment (eliska-n) [reply asab/library/providers/git.py]: Considering having this in abc - there is a different structure in the FileProvider... I'll think a bit about whether I could unify it somehow.

File: asab/library/providers/git.py (hunk 6 above)
Comment (eliska-n) [reply asab/library/providers/git.py]: I'm not publishing exact paths that have changed, but paths that the "user" has subscribed to and in which something happened, no matter what.

**Comment by eliska-n** [comment]
# TODO
### 2 race conditions:
- [ ] Someone might create a subfolder before I subscribe to a top directory - and the subfolder would remain without subscription. 
- [ ] https://github.com/TeskaLabs/asab/pull/331/files#diff-709a393caa5d278fd11b19ba5c05ab2f4ce8f6b668e5c6842450720cb2c30173R140

### 2 unhandled errors
- [ ] https://github.com/TeskaLabs/asab/pull/331/files#diff-709a393caa5d278fd11b19ba5c05ab2f4ce8f6b668e5c6842450720cb2c30173R158
- [ ] subscribing to a directory that does not exists raises KeyError (this one: https://github.com/TeskaLabs/asab/pull/331/files#diff-709a393caa5d278fd11b19ba5c05ab2f4ce8f6b668e5c6842450720cb2c30173R89) but it should be silent

### missing cleanup
- [ ] On every subscribe, `wd` is saved into `self.WDs`. But when the directory attached to this wd is removed, this reference remains in `self.WDs`. It is unclear to me what "removed" really means - the directory can be (at least) deleted or moved. I need to test it and find the right masks before I set some cleanup

File: asab/library/providers/library_utils/inotify.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
+import ctypes
+import struct
+
+libc6 = ctypes.cdll.LoadLibrary('libc.so.6')
```
Comment (ateska) [inline asab/library/providers/library_utils/inotify.py:4]: The name should be simplier: `filesystem_inotify.py` (eg.)

File: asab/library/service.py [hunk 8]
```diff
@@ -366,3 +366,15 @@ async def export(self, path="/", tenant=None):
 		tarobj.close()
 		fileobj.seek(0)
 		return fileobj
+
+	def subscribe(self, path):
+		# Normalize path
```
Comment (ateska) [inline asab/library/service.py:371]: Do asserts here to enforce user code to follow proper path strategy.

**Review by ateska — CHANGES_REQUESTED** [review]

File: asab/library/providers/filesystem.py [hunk 9]
```diff
@@ -34,6 +36,14 @@ def __init__(self, library, path, *, set_ready=True):
 		if set_ready:
 			self.App.TaskService.schedule(self._set_ready())
 
+		# Open inotify file descriptor
+		self.fd = inotify_init()
```
Comment (ateska) [inline asab/library/providers/filesystem.py:40]: `self.FD =`

File: asab/library/providers/filesystem_inotify.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+inotify_init = libc6.inotify_init
+inotify_init.argtypes = []
+inotify_init.restype = ctypes.c_int
+
+inotify_add_watch = libc6.inotify_add_watch
+inotify_add_watch.argtypes = [ctypes.c_int, ctypes.c_char_p, ctypes.c_uint32]
+inotify_add_watch.restype = ctypes.c_int
+
+
+_EVENT_FMT = 'iIII'
```
Comment (ateska) [inline asab/library/providers/filesystem_inotify.py:43]: `EVENT_FMT`

File: asab/library/providers/filesystem_inotify.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+_EVENT_SIZE = struct.calcsize(_EVENT_FMT)
```
Comment (ateska) [inline asab/library/providers/filesystem_inotify.py:44]: `EVENT_SIZE`

**Review by ateska — CHANGES_REQUESTED** [review]
Renaming.

File: asab/library/service.py [hunk 12]
```diff
@@ -366,3 +366,14 @@ async def export(self, path="/", tenant=None):
 		tarobj.close()
 		fileobj.seek(0)
 		return fileobj
+
+	def subscribe(self, paths):
```
Comment (eliska-n) [inline asab/library/service.py:370]: Subscribing list of paths - it makes sense to me when most of the subscriptions happen during init time and may require more than one path

File: asab/library/providers/filesystem.py [hunk 13]
```diff
@@ -102,3 +117,71 @@ async def list(self, path: str) -> list:
...
+		self.AggrEvents.clear()
+
+		for path in to_advertise:
+			if path is None:
+				continue
+			self.App.PubSub.publish("ASABLibrary.change!", self, path)
+
+
+	def subscribe(self, path):
+		if not os.path.isdir(self.BasePath + path):
```
Comment (eliska-n) [inline asab/library/providers/filesystem.py:162]: subscribing to nonexisting directory (or file) should lead to silent error

**Review by ateska — APPROVED** [review]

**Pull request merged**